File: sol/solution/0201-0300/0202/Solution.scala
object Solution {
  def isHappy(n: Int): Boolean = {
    def getNext(num: Int): Int =
      num.toString.map(c => (c.asDigit) * (c.asDigit)).sum
    var slow = n
    var fast = getNext(n)
    while (fast != 1 && slow != fast) {
      slow = getNext(slow)
      fast = getNext(getNext(fast))
    }
    fast == 1
  }
}

File: sol/solution/0201-0300/0202/Solution.swift
class Solution {
    func isHappy(_ n: Int) -> Bool {
        func getNext(_ num: Int) -> Int {
            var sum = 0, n = num
            while n > 0 {
                let d = n % 10
                sum += d * d
                n /= 10
            }
            return sum
        }
        var slow = n, fast = getNext(n)
        while fast != 1 && slow != fast {
            slow = getNext(slow)
            fast = getNext(getNext(fast))
        }
        return fast == 1
    }
}
